hir2/src/lib.rs

```diff
@@ -0,0 +1,61 @@
+#![feature(allocator_api)]
+#![feature(alloc_layout_extra)]
+#![feature(coerce_unsized)]
+#![feature(unsize)]
+#![feature(ptr_metadata)]
+#![feature(layout_for_ptr)]
+#![feature(slice_ptr_get)]
+#![feature(specialization)]
+#![feature(rustc_attrs)]
+#![feature(debug_closure_helpers)]
+#![feature(trait_alias)]
+#![feature(is_none_or)]
+#![feature(try_trait_v2)]
+#![feature(try_trait_v2_residual)]
+#![feature(tuple_trait)]
+#![feature(fn_traits)]
+#![feature(unboxed_closures)]
+#![feature(const_type_id)]
+#![feature(exact_size_is_empty)]
```

Comment:
There goes my hope to someday ditch the nightly. :)

Reply:
Yeah I've really blown a hole in that plan 😂. Once you need some of the nigh perma-unstable features like `ptr_metadata`, `fn_traits`, or `specialization`/`min_specialization`, it's basically game over for using stable, unless you have alternative options that allow you to sidestep the use of those features.

That said, a fair number of these are pretty close to stabilization. The ones that are problematic are:

* `specialization` and `rustc_attrs` are needed because `min_specialization` isn't _quite_ sufficient for the type-level magic I'm doing for automatically deriving verifiers for operations. If improvements are made to `min_specialization`, we can at least switch over to that. Longer term, we'd have to come up with an alternative approach to verification if we wanted to switch to stable though.
* `ptr_metadata` has an unclear stabilization path, and while it is basically table stakes for implementing custom smart pointer types, stabilizing it will require Rust to commit to some implementation details they are still unsure they want to commit to.
* `allocator_api` and `alloc_layout_extra` should've been stabilized ages ago IMO, but they are still being fiddled with. However, we aren't really leaning on these very much, and may actually be able to remove them.
* `coerce_unsized` and `unsize` are unlikely to be stabilized any time soon, as they've been a source of unsoundness in the language (not due to the features, but due to the traits themselves, which in stable are automatically derived). Unfortunately, they are also table stakes in implementing smart pointer types with ergonomics like the ones in libstd/liballoc, so its hard to avoid them.
* `try_trait_v2` and `try_trait_v2_residual` are purely ergonomic, so we could remove them in exchange for more verbose code in a couple places, not a big deal
* `fn_traits`, `unboxed_closures`, and `tuple_trait` are all intertwined with the ability to implement implementations of `Fn`/`FnMut`/`FnOnce` by hand, which I'm currently using to make the ergonomics of op builders as pleasant as possible. We might be able to tackle ergonomics in another way though, and as a result, no longer require these three features.

So, long story short, most of these provide some really significant ergonomics benefits - but if we _really_ wanted to ditch nightly for some reason, we could sacrifice some of those benefits and likely escape the need for most of these features. Verification is the main one that is not so easily abandoned, as the alternative is a lot of verbose and fragile boilerplate for verifying each individual operation - that's not to say it can't be done, just that I'm not sure the tradeoff is worth it.

Reply:
I can live with it. The verification is totally worth it!